src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs

```diff
@@ -0,0 +1,156 @@
+    }
+
+    fn visit_logical_values(&mut self, plan: &plan_node::LogicalValues) -> Cardinality {
+        plan.rows().len().into()
+    }
+
+    fn visit_logical_agg(&mut self, plan: &plan_node::LogicalAgg) -> Cardinality {
+        let input = self.visit(plan.input());
+
+        if plan.group_key().is_empty() {
+            input.min(1)
+        } else {
+            input.min(1..)
+        }
+    }
+
+    fn visit_logical_limit(&mut self, plan: &plan_node::LogicalLimit) -> Cardinality {
+        self.visit(plan.input()).min(plan.limit() as usize)
+    }
+
+    fn visit_logical_project(&mut self, plan: &plan_node::LogicalProject) -> Cardinality {
+        self.visit(plan.input())
+    }
+
+    fn visit_logical_top_n(&mut self, plan: &plan_node::LogicalTopN) -> Cardinality {
+        let input = self.visit(plan.input());
+
+        match plan.limit_attr() {
+            Limit::Simple(limit) => input.sub(plan.offset() as usize).min(limit as usize),
+            Limit::WithTies(limit) => {
+                assert_eq!(plan.offset(), 0, "ties with offset is not supported yet");
+                input.min((limit as usize)..)
+            }
+        }
+    }
+
+    fn visit_logical_filter(&mut self, plan: &plan_node::LogicalFilter) -> Cardinality {
+        let input = plan.input();
+        let eq_set: HashSet<_> = plan
+            .predicate()
+            .collect_input_refs(input.schema().len())
+            .ones()
+            .collect();
+
+        let mut unique_keys: Vec<HashSet<_>> = vec![input.logical_pk().iter().copied().collect()];
+
+        // We don't have UNIQUE key now. So we hack here to support some complex queries on
+        // system tables.
+        if let Some(scan) = input.as_logical_scan() && scan.is_sys_table() && scan.table_name() == PG_NAMESPACE_TABLE_NAME {
+            let nspname = scan.output_col_idx().iter().find(|i| scan.table_desc().columns[**i].name == "nspname").unwrap();
+            unique_keys.push([*nspname].into_iter().collect());
+        }
+
+        let input = self.visit(input);
+        if unique_keys
+            .iter()
+            .any(|unique_key| eq_set.is_superset(unique_key))
+        {
+            input.min(0..=1)
+        } else {
+            input.min(0..)
+        }
+    }
+
+    fn visit_logical_union(&mut self, plan: &plan_node::LogicalUnion) -> Cardinality {
+        let all = plan
+            .inputs()
+            .into_iter()
+            .map(|input| self.visit(input))
+            .fold(Cardinality::default(), std::ops::Add::add);
+
+        if plan.all() {
+            all
+        } else {
+            all.min(1..)
+        }
+    }
+
+    fn visit_logical_join(&mut self, plan: &plan_node::LogicalJoin) -> Cardinality {
+        let left = self.visit(plan.left());
+        let right = self.visit(plan.right());
+
+        match plan.join_type() {
+            JoinType::Unspecified => unreachable!(),
+
+            JoinType::Inner => left.mul(right.min(0..)),
```

Comment:
Can you add comments like this to help understand 

![FC3EBFF7-5CAC-4ED5-9CDE-97DF3006CB96](https://user-images.githubusercontent.com/37948597/233294039-eeb82af3-33e6-44f9-9112-fc62b5041322.png)